pkg/provider/ngrok/client.go

```diff
@@ -0,0 +1,329 @@
+/*
+Copyright © 2025 ESO Maintainer Team
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ngrok
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ngrok/ngrok-api-go/v7"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/utils/ptr"
+
+	esv1 "github.com/external-secrets/external-secrets/apis/externalsecrets/v1"
+	"github.com/external-secrets/external-secrets/pkg/utils/metadata"
+)
+
+const (
+	defaultDescription = "Managed by External Secrets Operator"
+)
+
+var (
+	errWriteOnlyOperations     = errors.New("not implemented - the ngrok provider only supports write operations")
+	errVaultDoesNotExist       = errors.New("vault does not exist")
+	errVaultSecretDoesNotExist = errors.New("vault secret does not exist")
+	errCannotPushNilSecret     = errors.New("cannot push nil secret")
+)
+
+type PushSecretMetadataSpec struct {
+	// The description of the secret in the ngrok API.
+	Description string `json:"description,omitempty"`
+	// Custom metadata to be merged with generated metadata for the secret in the ngrok API.
+	// This metadata is different from Kubernetes metadata.
+	Metadata map[string]string `json:"metadata,omitempty"`
+}
+
+type VaultClient interface {
+	Create(context.Context, *ngrok.VaultCreate) (*ngrok.Vault, error)
+	Get(context.Context, string) (*ngrok.Vault, error)
+	GetSecretsByVault(string, *ngrok.Paging) ngrok.Iter[*ngrok.Secret]
+	List(*ngrok.Paging) ngrok.Iter[*ngrok.Vault]
+}
+
+type SecretsClient interface {
+	Create(context.Context, *ngrok.SecretCreate) (*ngrok.Secret, error)
+	Delete(context.Context, string) error
+	Get(context.Context, string) (*ngrok.Secret, error)
+	List(*ngrok.Paging) ngrok.Iter[*ngrok.Secret]
+	Update(context.Context, *ngrok.SecretUpdate) (*ngrok.Secret, error)
+}
+
+type client struct {
+	vaultClient   VaultClient
+	secretsClient SecretsClient
+	vaultName     string
+	vaultID       string
+	vaultIDMu     sync.RWMutex
+}
+
+func (c *client) PushSecret(ctx context.Context, secret *corev1.Secret, data esv1.PushSecretData) error {
+	if secret == nil {
```

Comment:
Did you ever actually run into this? I believe the contract implies you will not be called with a nil secret.

Reply:
PushSecret is called from a for loop of secrets. If there would be no secret it wouldn't be called. :)

Reply:
Is it okay to leave this validation or do I need to remove it to resolve this?

Reply:
Removed in https://github.com/external-secrets/external-secrets/pull/5160/commits/2f826f6dbed18f8752b7ac2374318e1cc5558d44